src/sentry/utils/snuba.py

```diff
@@ -668,21 +687,23 @@ def bulk_raw_query(
     if referrer:
         headers["referer"] = referrer
 
+    def _prep_query(query: Union[Query, SnubaQueryParams]) -> SnubaQuery:
+        if isinstance(query, Query):
+            return query, lambda x: x, lambda x: x
```

Comment:
Why don't you move this logic into `raw_snql_query` and also pass a `get_cache_key` function to use in here ? IT does not seem useless to accept a Union. Once you are past line 704 there is no difference between the two implementation. This was you would not have at the same time the SnQL only path and the path where you run the old query and the SnQL one in parallel mixed up in the same method.

Reply:
I'm not sure I understand what you're suggesting. I can move the SnQL part of `_prep_query` to `raw_snql_query` but I need to keep `_prepare_query_params` in `bulk_raw_query` because `bulk_raw_query` is called directly sometimes. If that's the case, then I need to have an if check here.

As for passing the cache function, what's wrong with what I have?

Reply:
> bulk_raw_query is called directly sometimes

Is a product team supposed to use only `raw_snql_query` to build a new feature based on SnQL or is there a reason to still use `bulk_raw_query`. I was under the impression that a new feature would only use `raw_snql_query`.
If we expect clients to call `bulk_raw_query` directly with a SnQL query, why do we need a public `raw_snql_query` method ?

My comment above was assuming we would have two separate public entry points, one for legacy queries and another one for SnQL, In that case we should not need a single method (bulk_raw_query) to take a Union. The SnQL method could adapt the query into a SnubaQuery and then depend on the raw_bulk method that now has one code path less.

Reply:
thanks, sorry for the delay in the answer